accounts/ipc.py

```diff
@@ -0,0 +1,83 @@
+from datetime import timedelta
+
+import requests
+from django.utils import timezone
+
+from accounts.settings import accounts_settings
+
+
+def authenticated_request(user, method, url,
+                          params=None, data=None, headers=None, cookies=None, files=None,
+                          auth=None, timeout=None, allow_redirects=True, proxies=None,
+                          hooks=None, stream=None, verify=None, cert=None, json=None):
+    """
+    Helper method to make an authenticated request using the user's access token
+    NOTE be ABSOLUTELY sure you only make a request to Penn Labs products, otherwise
+    you will expose user's access tokens to the URL you provide and bad things will
+    happen
+    """
+
+    # Access token is expired. Try to refresh access token
+    if user.accesstoken.expires_at < timezone.now():
+        if not refresh_access_token(user):
+            return None  # Couldn't update access token
+
+    # Update Headers
+    headers = headers or {}
+    headers['Authorization'] = f'Bearer {user.accesstoken.token}'
+
+    # Make request
+    s = requests.Session()
+    return s.request(
+        method=method,
+        url=url,
+        params=params,
+        data=data,
+        headers=headers,
+        cookies=cookies,
+        files=files,
+        auth=None,
+        timeout=None,
+        allow_redirects=allow_redirects,
+        proxies=proxies,
+        hooks=hooks,
+        stream=stream,
+        verify=verify,
+        cert=cert,
+        json=json,
+    )
+
+
+def refresh_access_token(user):
+    """
+    Helper method to update a user's access token. Should be used when a user's
+    access token has expired, but still has a valid refresh token.
+
+    Returns:
+        bool: true if the access token is updated, false otherwise.
+    """
+    body = {
+        'grant_type': 'refresh_token',
+        'client_id': accounts_settings.CLIENT_ID,
+        'client_secret': accounts_settings.CLIENT_SECRET,
+        'refresh_token': user.refreshtoken.token,
+    }
+    try:
+        data = requests.post(
+            url=accounts_settings.PLATFORM_URL + '/accounts/token/',
+            data=body
+        )
+        if data.status_code == 200:  # Access token refreshed successfully
+            data = data.json()
+            # Update Access token
+            user.accesstoken.token = data['access_token']
+            user.accesstoken.expires_at = timezone.now() + timedelta(seconds=data['expires_in'])
+            user.accesstoken.save()
+
+            # Update Refresh Token
+            user.refreshtoken.token = data['refresh_token']
+            user.refreshtoken.save()
+
+    except requests.exceptions.RequestException:  # Can't connect to platform
+        return False
```

Comment:
If the caller method throws a 403 when platform can't be reached, I think this function should also throw a 403 when platform can't be reached. Or a 502 in both places if you wanna do that ;)

Reply:
This method doesn't return an HTTP response, so it wouldn't really be possible to cleanly return a 403 as well. This is also supposed to be a strictly internal function as well, so I'm less worried about what it returns, the actual logic for what to do when you can't reach platform exists in the `authenticated_request` method. I can probably prefix this method with an underscore to signify that too.

Reply:
Pretty sure you throw 403s rather than return them. Honestly think it's fine, considering then it's just an exception that can be caught like any other. If you'd like to keep this non-http though, then try to throw a 403 from the main function if this returns false.

Reply:
I don't think we should throw any kind of exception if the access token can't be updated. If we do that, then either the view trying to make the authenticated request will have to handle the exception, or the entire view breaks. This is something where it makes more sense for it to silently fail to me. Maybe instead, we can return a response object with a hardcoded 403 status code so that the view trying to make the request can handle it the same way as if the user didn't have access to the desired endpoint.

Reply:
Fixed